include/ghex/structured/cubed_sphere/halo_generator.hpp

```diff
@@ -0,0 +1,281 @@
+/* 
+ * GridTools
+ * 
+ * Copyright (c) 2014-2020, ETH Zurich
+ * All rights reserved.
+ * 
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * 
+ */
+#ifndef INCLUDED_GHEX_STRUCTURED_CUBED_SPHERE_HALO_GENERATOR_HPP
+#define INCLUDED_GHEX_STRUCTURED_CUBED_SPHERE_HALO_GENERATOR_HPP
+
+#include <algorithm>
+#include <vector>
+#include "./transform.hpp"
+#include "./domain_descriptor.hpp"
+
+namespace gridtools {
+    namespace ghex {
+        namespace structured {
+            namespace cubed_sphere {
+        
+                /** @brief halo generator for cubed-sphere grids */
+                class halo_generator {
+                public: // member types
+                    using domain_type = domain_descriptor;
+                    using coordinate_type = domain_type::coordinate_type;
+
+                private: // member types
+                    // describes an iteration space
+                    struct box {
+                        const coordinate_type& first() const { return m_first; }
+                        const coordinate_type& last() const { return m_last; }
+                        coordinate_type& first() { return m_first; }
+                        coordinate_type& last() { return m_last; }
+                        coordinate_type m_first;
+                        coordinate_type m_last;
+                    };
+
+                    // describes a pair of iteration spaces (local and global)
+                    struct box2 {
+                        const box& local() const { return m_local; }
+                        const box& global() const { return m_global; }
+                        box& local() { return m_local; }
+                        box& global() { return m_global; }
+                        box m_local;
+                        box m_global;
+                    };
+
+                private: // members
+                    std::array<int, 4> m_halo;
+
+                public: // ctors
+                    /** @brief add an equal halo around the domain (default case)
+                      * @param halo number of halo cells */
+                    halo_generator(int halo) noexcept : m_halo{halo,halo,halo,halo} {}
+
+                    /** @brief construct with custom halo
+                      * @tparam Array 4-D array type
+                      * @param halos halos in +x,-x,-y,+y directions (for staggered grids) */
```

Comment:
for staggered grids you might want to use different halos on each side of the grid.
In general, for unstructured grids (which this kind of is when projected to the sphere) it makes sense to just say how many neighbour cells we should consider (i.e. 1 number). In the special case of cubed-sphere grids, it can be useful to specify halos per direction to easily enable staggering.

Reply:
Right, I see